torch/autograd/functional.py

```diff
@@ -0,0 +1,714 @@
+            a differentiable manner. Note that when strict if ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect we detect that some quantities
+            are independent of the element we want to differentiate it with. If ``False``, such quantity
+            will only be set to ``0`` which is the expected mathematical value.
+            Defaults to ``False``.
+
+    Returns:
+        Jacobian (Tensor or nested tuple of Tensor) if there are a single input
+            and output, this will be a single Tensor containing the Jacobian for the
+            linearized inputs and output. If one of the two is a tuple, then the Jacobian
+            will be a tuple of Tensors. If both of them are tuples, then the Jacobian will
+            be a tuple of tuple of Tensors where ``Jacobian[i][j]`` will contain the Jacobian
+            of the ``i``th output and ``j``th input and will have as size the sum of the size
+            of the corresponding output plus the size of the corresponding input.
+
+    Example::
+
+        >>> def exp_reducer(x):
+        ...   return x.exp().sum(dim=1)
+        >>> inputs = torch.rand(2, 2)
+        >>> jacobian(exp_reducer, inputs)
+        tensor([[[1.4917, 2.4352],
+                 [0.0000, 0.0000]],
+
+                [[0.0000, 0.0000],
+                 [2.4369, 2.3799]]])
+
+        >>> jacobian(exp_reducer, inputs, create_graph=True)
+        tensor([[[1.4917, 2.4352],
+                 [0.0000, 0.0000]],
+
+                [[0.0000, 0.0000],
+                 [2.4369, 2.3799]]], grad_fn=<ViewBackward>)
+
+        >>> def exp_adder(x, y):
+        ...   return 2 * x.exp() + 3 * y
+        >>> inputs = (torch.rand(2), torch.rand(2))
+        >>> jacobian(exp_adder, inputs)
+        (tensor([[2.8052, 0.0000],
+                [0.0000, 3.3963]]),
+         tensor([[3., 0.],
+                 [0., 3.]]))
+    """
+
+    tuple_inputs, inputs = _as_tuple(inputs, "inputs", "jacobian")
+    inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
+
+    outputs = func(*inputs)
+    tuple_outputs, outputs = _as_tuple(outputs, "outputs of the user-provided function", "jacobian")
+    _check_requires_grad(outputs, "outputs", strict=strict)
+
+    jacobian = tuple()
+    for i, out in enumerate(outputs):
+
+        jac_i = tuple([] for _ in range(len(inputs)))
+        for j in range(out.nelement()):
+            # Do this instead of the grad_output trick in `gradcheck` to avoid version counter issues
+            vj = _autograd_grad(out.reshape(-1)[j], inputs, retain_graph=True, create_graph=create_graph)
+
+            for el_idx, (jac_i_el, vj_el, inp_el) in enumerate(zip(jac_i, vj, inputs)):
+                # Note that grad_out_view might share memory with vj
```

Comment:
What is "grad_out_view"?

Reply:
Ho that's an old comment from when I was using the `grad_ouptut` trick mentionned above (in the version counter note).
This comment, and the `.clone()` on `vj_el` below can be removed. Good catch.